http-gateways/DNSSEC_EXPORT_DNSLINK.md

```diff
@@ -0,0 +1,313 @@
+
+This specification consists of four logical pieces.
+
+1. A method for serializing DNSSEC proofs and IPFS CAR files
+2. A method for communicating these proofs to the requesting client
+3. A method for the client to validate DNSSEC proofs
+4. A method for the client to validate the integrity of the IPFS content
+
+## Motivation
+In the current gateway, clients are forced to trust the DNS and IPNS resolutions of the gateway. This entails implicit risk for clients who cannot validate if the content they receive is the content bound to the name they requested. This state of affairs can leave clients vulnerable to MitM attacks.
+
+This proposal reduces the trust needed by a client, and allows honest gateways to prove it is responding truthfully.
+
+## Requirements & Assumptions
+### Feature Requirements
+  1. The proof must be returned in-band with the DNS responses so that the gateway does not need to store state. This is not a hard requirement, but storing state is an engineering challenge that would be nice to avoid.
+  2. The proof must be verifiable by a client. This includes not being so large as to cause undue burden on the client.
+
+### Compatibility Requirements
+1. Participating clients should still be compatible with non-participating gateways. In other words, a client benefiting from this solution should still be compatible with all gateways that it can presently use.
+2. Gateways should maintain compatibility with non-participating clients—those clients will not benefit from the DNSSEC proofs, but they should still have full access to data.
+3. Gateways should not rely on any specific public DNS resolver, i.e., the implementation should be resolver agnostic.
+
+### Security Considerations
+1. DNSSEC responses can be much larger that vanilla DNS responses, in theory, this could be a vector for DDoS amplification.
+
+### Assumptions on Scope
+1. We consider only IPFS blocks bound to DNS with DNSLink. This is without loss of generality, because an IPFS subgraph referenced by its root hash is directly verifiable by confirming the hashes and thus would not benefit from the DNSSEC proof.
+2. For non-existent URLs, we will not walk the zone (through NSEC records). Including the walkthrough would dramatically increase the proof size. In other words, we will only offer verifiable proof for positive responses.
+3. We will not support wildcards in DNS resolution. This is not a problem because our target subdomains will always begin with "\_dnslink", and domains cannot have wildcards in the middle.
+4. We will not support multiple root Key Signing Keys or changed Key Signing Keys in this version.
```

Comment:
I wrote that because if the root KSK changes, then whatever client-side program is doing the DNSSEC validation would need to be updated out of band to use the new KSK.

But you're right that I generally assumed that the client is aware of the current KSK situation. I think it will be more clear if I remove \#4 and replace it with an explicit statement about the client's awareness of the root's KSK situation.

```suggestion
4. We assume that the client is configured to use the correct root zone KSK.
```